File: Java/src/main/java/sort/QuickSort.java
package sort;


public class QuickSort {

    public static void quickSort(int[] arr){
        quickSortBetween(arr,0,arr.length-1);
    }

    private static  void quickSortBetween(int[] arr,int low,int high){
        if (low>=high)return ;
        int pivot = partition(arr,low,high);
        quickSortBetween(arr,low,pivot-1);
        quickSortBetween(arr,pivot+1,high);
    }

    private  static  int partition(int[] arr,int low,int high){
        int j = low ;
        int pivot = arr[low];
        int i = low + 1 ;
        while (i <= high){
            if (arr[i]<arr[low]){
                j ++ ;
                int tmp = arr[i];
                arr[i] =arr[j];
                arr[j] = tmp;
            }
            i++;
        }
        int tmp = arr[j];
        arr[j] =arr[low];
        arr[low] = tmp;
        return j;
    }

    public static void main(String[] args) {
        int[] arr = {3,2,14,6,1};
        quickSort(arr);
        for (int i:
             arr) {
            System.out.print(i+",");
        }
    }

}

File: Python/algorithm/sort/QuickSort.py
"""
快速排序

递归公式：quick_sort(s,e) = partition->m  qucik_sort(s,m-1) + quick_sort(m+1,e)
终止条件：s>=e
"""
from typing import List


def quickSort(_arr:List[int]):
    _quickSortBetween(_arr,0,len(_arr)-1)

def _quickSortBetween(_arr:List[int], _low:int, _high:int):
    if _low >= _high:
        return
    pivot = _partition(_arr,_low,_high)
    _quickSortBetween(_arr,_low,pivot-1)
    _quickSortBetween(_arr,pivot+1,_high)

def _partition(_arr:List[int], _low:int, _high:int) -> int:
    j = _low
    pivot = _arr[_low]
    for i in range(_low+1,_high+1):
        if _arr[i]<=pivot:
            j += 1
            _arr[i],_arr[j] = _arr[j],_arr[i]
    _arr[_low],_arr[j] = _arr[j],_arr[_low]
    return j
if __name__ == '__main__':
    arr = [3,4,2,1,6]
    quickSort(arr)
    print(arr)